zilencer: Register push device to bouncer.

prakhar1144 opened the pull request:
API Design threads:
* [#api design > E2EE - register push device to bouncer](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20-.20register.20push.20device.20to.20bouncer/with/2199387)
* [#api design > E2EE - error raised by bouncer](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20-.20error.20raised.20by.20bouncer/with/2199792)

Backend thread:
* [#backend > E2EE - asymmetric key-pair on bouncer](https://chat.zulip.org/#narrow/channel/3-backend/topic/E2EE.20-.20asymmetric.20key-pair.20on.20bouncer/with/2188685)

ZAP: https://github.com/zulip/zulip-architecture/pull/2



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zilencer/migrations/0065_remotepushdevice.py:
```diff
@@ -0,0 +1,39 @@
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("zerver", "0718_fix_topics_for_direct_messages"),
```
Does this actually need a dependency on `zerver` at all? I guess you have the nullable `realm` field, and that's why? This is OK for now but may need refactoring when we next squash migrations; work can deferred until then I expect.

prakhar1144 replied:
Yeah,

For self-hosters: `remote_realm` will be set and `realm` set to null.
For cloud: `realm` will be set and `remote_realm` set to null.

timabbott commented on zilencer/migrations/0065_remotepushdevice.py:
```diff
@@ -0,0 +1,39 @@
+            fields=[
+                ("device_id", models.BigAutoField(primary_key=True, serialize=False)),
+                ("token", models.TextField()),
+                ("token_kind", models.PositiveSmallIntegerField(choices=[(1, "Apns"), (2, "Fcm")])),
```
The capitalization here is pretty atypical. I think we should use standard spellings of these like we do in the `VisibilityPolicy` declaration.

prakhar1144 replied:
Changed it to:
```py
    class TokenKind(models.IntegerChoices):
        APNS = 1, "APNs"
        FCM = 2, "FCM"

    token_kind = models.PositiveSmallIntegerField(choices=TokenKind.choices)
```

timabbott commented on zilencer/models.py:
```diff
@@ -587,3 +587,26 @@ def has_stale_audit_log(server: RemoteZulipServer) -> bool:
     return False
+
+
+class RemotePushDevice(models.Model):
```
This should get comments for the fields, like one before realm/remote_realm explaining why we have both columns.

prakhar1144 replied:
Comments added.

timabbott commented on scripts/setup/generate_secrets.py:
```diff
@@ -180,6 +184,20 @@ def add_secret(name: str, value: str) -> None:
+            private_key_str = Base64Encoder.encode(bytes(private_key)).decode("utf-8")
+            public_key_str = Base64Encoder.encode(bytes(private_key.public_key)).decode("utf-8")
+            add_secret(
+                "push_registration_encryption_keys", json.dumps({public_key_str: private_key_str})
```
This JSON pattern inside a single secret is a bit unusual; is there a reason to prefer it to a pair of settings like we have with (`zulip_org_id`/`zulip_org_key`)? 

I think the reason folks were talking about a JSON file is they were imagining a separate file off the side; but it probably does make sense to put this in the normal secrets file, in which case I think just two secrets is reasonable.

prakhar1144 replied:
> This JSON pattern inside a single secret is a bit unusual; is there a reason to prefer it to a pair of settings like we have with (zulip_org_id/zulip_org_key)?

We plan to support multiple public-private key pair at once. 

The ZAP says:
> The push bouncer maintains a push_registration_encryption_keys map containing asymmetric key pairs ({public_key: private_key}), each generated to support key rotation. When a new key pair is created, the public key (bouncer_public_key) is baked into a new client release.

> To support older clients, the bouncer continues to accept the older public keys for a period of time, allowing clients to upgrade. Once a sufficient portion of clients have updated, the corresponding old keys are removed from the map.

timabbott replied:
Yeah that seems fine for now. Easy to migrate later, since this is zilencer side configuration only.

timabbott commented on scripts/setup/generate_secrets.py:
```diff
@@ -180,6 +184,20 @@ def add_secret(name: str, value: str) -> None:
     if need_secret("zulip_org_id"):
         add_secret("zulip_org_id", str(uuid.uuid4()))
 
+    if need_secret("push_registration_encryption_keys"):
```
Will this result generating these on any Zulip server, or just on the bouncer server? I think `need_secret` would generate this on every server, but I wasn't sure if that was intended.

timabbott replied:
OK, the thing I was missing is that we may need these for a development environment. We're never going to "generate secrets during installation" for Zulip Cloud, which is the production deployment of this system.

I think you can have this code check the `development` and only generate this secret there and that'll resolve this thread.

prakhar1144 replied:
Made the changes:
```py
    if development and need_secret("push_registration_encryption_keys"):
        private_key = PrivateKey.generate()
        private_key_str = Base64Encoder.encode(bytes(private_key)).decode("utf-8")
        public_key_str = Base64Encoder.encode(bytes(private_key.public_key)).decode("utf-8")
        add_secret(
            "push_registration_encryption_keys", json.dumps({public_key_str: private_key_str})
        )
```

timabbott replied:
OK. Can you also add a comment:

> settings.ZILENCER_ENABLED would be a better check than development for whether we need push bouncer secrets, but we're trying to avoid importing settings.

prakhar1144 replied:
Comment added.

timabbott commented on zilencer/views.py:
```diff
@@ -475,6 +487,102 @@ def register_remote_push_device(
+    request: HttpRequest,
+    server: RemoteZulipServer,
+    *,
+    user_uuid: str,
```
I don't think `user_uuid` is present in this protocol?

prakhar1144 replied:
`user_uuid` is required by `get_remote_realm_helper()` for logging purpose. 

It's not included in the ZAP, yeah. -- [CZO Discussion](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20-.20register.20push.20device.20to.20bouncer/near/2199387)

timabbott replied:
OK replied in the chat.zulip.org thread, but I think the answer is very likely to be that we should drop this parameter; the main question is how to adjust `get_remote_realm_helper` -- do we make `user_uuid` nullable or remove it entirely.

timabbott commented on zilencer/views.py:
```diff
@@ -475,6 +487,102 @@ def register_remote_push_device(
+    except PushRegistrationDecodeError:
+        raise InvalidEncryptedPushRegistrationError
+    except PushRegistrationLifespanExceededError:
+        raise PushRegistrationLivenessTimedOutError
```
Is there a reason you have this set of pairs of exception classes? I'd expect you could just have `do_register_remote_push_device` raise the right exception in the first place.

timabbott replied:
I see. I don't see how this approach is any less messy than just catching the relevant JsonableError subclass exceptions in the Zulip Cloud code path instead. Extra fields/methods are basically free, so there's no real performance impact.

timabbott replied:
Yeah there's no reason to logically separate that. The purpose of `JsonableError` is to be an exception which *can* be converted into JSON and will do that if not caught, but also permits code to catch the exception to handle it where appropriate.

timabbott commented on zilencer/views.py:
```diff
@@ -475,6 +487,102 @@ def register_remote_push_device(
+            else None
+        )
+        timestamp = push_registration["timestamp"]
+    except (TypeError, CryptoError, orjson.JSONDecodeError, KeyError, JsonableError):
```
It's a better pattern to have a series of smaller try/except blocks that each only catch the expected exception types.

prakhar1144 replied:
Made the changes to use smaller try/except blocks.

timabbott commented on zilencer/views.py:
```diff
@@ -475,6 +487,102 @@ def register_remote_push_device(
+        token = push_registration["token"]
+        token_kind = push_registration["token_kind"]
+        validate_token(token, token_kind)
+        # TODO: Add validation for ios_app_id and timestamp
```
Probably here you should be using Pydantic to parse the `push_registration` object, potentially including the JSON-decoding party. It's a very nice library that we already use for this purpose, and it can handle all the type validation consistently for us, giving a nice typed dataclass that the following code can work with.

prakhar1144 replied:
Didn't get a chance to use Pydantic in depth previously, reworked this to use Pydantic. Looks like an awesome library.

timabbott replied:
Yeah it's great :)

timabbott commented on zerver/tests/test_push_notifications.py:
```diff
@@ -1342,6 +1345,108 @@ def test_push_bouncer_api(self) -> None:
+            self.server_uuid,
+            "/api/v1/remotes/push/e2ee/register",
+            payload,
+        )
```
You should parse the response here and make sure it returned the right device ID too. 

Actually this test code needs to assert success too -- you never want to make an API call and then not check the result in tests, and it can hide nasty bugs :)

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments
* Extended test to verify more error conditions
* Unique constraint added

Posted a question in [#backend > uv @ 💬](https://chat.zulip.org/#narrow/channel/3-backend/topic/uv/near/2203179)

andersk commented on scripts/setup/generate_secrets.py:
```diff
@@ -1,9 +1,13 @@
 import sys
 from contextlib import suppress
 
+from nacl.encoding import Base64Encoder
+from nacl.public import PrivateKey
```
This needs to go after `setup_path()`.

andersk commented on zerver/lib/test_classes.py:
```diff
@@ -2284,16 +2284,17 @@ def fetch_other_streams_subscriber_count(self, stream_ids: set[int]) -> dict[int
             continue
-        ids = model._default_manager.all().values_list("id", flat=True)
-        yield table_name, set(ids)
+        pk_name = model._meta.pk.name
+        pks = model._default_manager.all().values_list(pk_name, flat=True)
```
```suggestion
        pks = model._default_manager.all().values_list("pk", flat=True)
```

prakhar1144 replied:
Done

andersk commented on zerver/tests/test_push_notifications.py:
```diff
@@ -1342,6 +1345,148 @@ def test_push_bouncer_api(self) -> None:
+        token: str = "apple-tokenaz",
+        token_kind: int = RemotePushDevice.TokenKind.APNS,
+        ios_app_id: str | None = "example.app",
+        timestamp: int = datetime_to_timestamp(now()),
```
Python defaults are evaluated at function definition time, so this gives the timestamp at which the function was defined, not the timestamp at which it was called.

timabbott replied:
If code that looks like it's buggy is intentional, it HAS to have a comment explaining what you're doing. Code that looks buggy is very expensive for people reading code, both during code review and for years into the future.

But once you're doing that, you almost always will decide it's better to write it the way that doesn't look like a bug.

prakhar1144 replied:
```diff
-        timestamp: int = datetime_to_timestamp(now()),
+        timestamp: int | None = None,
     ) -> dict[str, str | int]:
         hamlet = self.example_user("hamlet")
         remote_realm = RemoteRealm.objects.get(uuid=hamlet.realm.uuid)
 
+        if timestamp is None:
+            timestamp = datetime_to_timestamp(now())
+
```

timabbott commented on zilencer/models.py:
```diff
@@ -587,3 +587,42 @@ def has_stale_audit_log(server: RemoteZulipServer) -> bool:
+    class Meta:
+        constraints = [
+            UniqueConstraint(
+                fields=["push_account_id", "token"],
```
Yes this will be tiny compared to other tables. We should define what database indexes we want.

What is the purpose of this index, though? Probably a comment should document what failure mode it's meant to prevent.

timabbott commented on zilencer/models.py:
```diff
@@ -587,3 +587,42 @@ def has_stale_audit_log(server: RemoteZulipServer) -> bool:
+class RemotePushDevice(models.Model):
+    """Each row corresponds to an account on an install of the app
+    registered to receive mobile push notifications.
+    """
```
That's a good question. Posed in CZO: [#api design > E2EE - RemotePushDevice tab le @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/E2EE.20-.20RemotePushDevice.20tab.20le/near/2203895)

prakhar1144 commented on zilencer/models.py:
```diff
@@ -587,3 +587,42 @@ def has_stale_audit_log(server: RemoteZulipServer) -> bool:
+    remote_realm = models.ForeignKey(RemoteRealm, on_delete=models.CASCADE, null=True)
+
+    # A token that uniquely identifies the app instance.
+    token = models.CharField(max_length=4096)
```
The `CharField` results in `varchar(n)` and not `character(n)` type. So, it won't consume 4KB for each row.

Also, [TextField docs mentions](https://docs.djangoproject.com/en/5.2/ref/models/fields/#django.db.models.TextField):
> If you specify a max_length attribute, it will be reflected in the [Textarea](https://docs.djangoproject.com/en/5.2/ref/forms/widgets/#django.forms.Textarea) widget of the auto-generated form field. However it is not enforced at the model or database level. Use a [CharField](https://docs.djangoproject.com/en/5.2/ref/models/fields/#django.db.models.CharField) for that.

So, I think `CharField` is fine. Extended the comment:
```py
    # A token that uniquely identifies the app instance.
    #
    # FCM and APNs don't specify a maximum token length, so we only enforce
    # that they're at most the maximum FCM / APNs payload size of 4096 bytes.
    token = models.CharField(max_length=4096)
```

timabbott replied:
OK.

timabbott commented:
Merged prep commits from #34836 using this PR ID. Changes to the model definition were just comments:
``` diff
diff --git a/zilencer/models.py b/zilencer/models.py
index 1465617b0a..ac142dbb06 100644
--- a/zilencer/models.py
+++ b/zilencer/models.py
@@ -95,7 +95,12 @@ class RemoteZulipServer(models.Model):
 
 
 class RemotePushDeviceToken(AbstractPushDeviceToken):
-    """Like PushDeviceToken, but for a device connected to a remote server."""
+    """Class for storing legacy non-E2EE mobile push
+    tokens. Deprecated in favor of RemotePushDevice, which is designed
+    for E2EE mobile notifications.
+
+    Like PushDeviceToken, but for a device connected to a remote server.
+    """
 
     server = models.ForeignKey(RemoteZulipServer, on_delete=models.CASCADE)
     # The user id on the remote server for this device
@@ -590,19 +595,24 @@ def has_stale_audit_log(server: RemoteZulipServer) -> bool:
 
 
 class RemotePushDevice(models.Model):
-    """Each row corresponds to an account on an install of the app
+    """Core bouncer server table storing registrations to receive
+    mobile push notifications via the bouncer server.
+
+    Each row corresponds to an account on an install of the app
     registered to receive mobile push notifications.
     """
 
+    # Unique identifier assigned by the bouncer for this registration.
     device_id = models.BigAutoField(primary_key=True)
 
-    # Set to null for self-hosters. `remote_realm` is set instead.
+    # Realm that the account is associated with. Both Cloud and
+    # self-hosted are supported; exactly one of these will be null.
     realm = models.ForeignKey(Realm, on_delete=models.CASCADE, null=True)
-
-    # Set to null for zulip cloud. `realm` is set instead.
     remote_realm = models.ForeignKey(RemoteRealm, on_delete=models.CASCADE, null=True)
 
-    # A token that uniquely identifies the app instance.
+    # The token that uniquely identifies the app instance to the
+    # FCM/APNs servers. There will be duplicates in this table when a
+    # device is logged into multiple Zulip realms.
     #
     # FCM and APNs don't specify a maximum token length, so we only enforce
     # that they're at most the maximum FCM / APNs payload size of 4096 bytes.
@@ -614,12 +624,17 @@ class RemotePushDevice(models.Model):
 
     token_kind = models.PositiveSmallIntegerField(choices=TokenKind.choices)
 
-    # ID to identify an account within an install of the app.
+    # 64-bit random integer ID generated by the client; will only be
+    # guaranteed to be unique within the client's own table of accounts.
     push_account_id = models.BigIntegerField()
 
+    # If the token is expired, the date when the bouncer learned it
+    # was expired via an error from the FCM/APNs server. Used to
+    # support delayed deletion. Null if the bouncer believes this
+    # token is still valid.
     expired_time = models.DateTimeField(null=True)
 
-    # [optional] Contains the app id of the device if it is an iOS device
+    # Contains the app id of the device if and only if token_kind is APNs.
     ios_app_id = models.TextField(null=True)
 
     class Meta:
```

timabbott commented:
Oops, apparently my comment changes weren't included in the merge. Added in 71e8fcbb3a26bc27b7ec70b95434d6a03cf12549.
